Core/Src/main.c

```diff
@@ -0,0 +1,595 @@
+/* USER CODE BEGIN Header */
+/**
+  ******************************************************************************
+  * @file           : main.c
+  * @brief          : Main program body
+  ******************************************************************************
+  * @attention
+  *
+  * Copyright (c) 2023 STMicroelectronics.
+  * All rights reserved.
+  *
+  * This software is licensed under terms that can be found in the LICENSE file
+  * in the root directory of this software component.
+  * If no LICENSE file comes with this software, it is provided AS-IS.
+  *
+  ******************************************************************************
+  */
+/* USER CODE END Header */
+/* Includes ------------------------------------------------------------------*/
+#include "main.h"
+
+/* Private includes ----------------------------------------------------------*/
+/* USER CODE BEGIN Includes */
+#define DUTY_CYCLE 2000 //unused, can prolly delete
+/* USER CODE END Includes */
+
+/* Private typedef -----------------------------------------------------------*/
+/* USER CODE BEGIN PTD */
+
+/* USER CODE END PTD */
+
+/* Private define ------------------------------------------------------------*/
+/* USER CODE BEGIN PD */
+
+/* USER CODE END PD */
+
+/* Private macro -------------------------------------------------------------*/
+/* USER CODE BEGIN PM */
+
+/* USER CODE END PM */
+
+/* Private variables ---------------------------------------------------------*/
+TIM_HandleTypeDef htim2;
+
+UART_HandleTypeDef huart2;
+
+/* USER CODE BEGIN PV */
+
+/* USER CODE END PV */
+
+/* Private function prototypes -----------------------------------------------*/
+void SystemClock_Config(void);
+static void MX_GPIO_Init(void);
+static void MX_USART2_UART_Init(void);
+static void MX_TIM2_Init(void);
+/* USER CODE BEGIN PFP */
+
+/* USER CODE END PFP */
+
+/* Private user code ---------------------------------------------------------*/
+/* USER CODE BEGIN 0 */
```

Comment:
I mentioned a few times to add a set of variables to their own struct.

You should do something like this:

```
typedef struct
{
    int PropTerm = 10; //proportional term (30?)
    float IntTerm = 0; //Integral term
    float DerivTerm = 0.1; //Derivative term
} PID_CONTROL_GAINS;

typedef struct
{
    PID_CONTROL_GAINS gains;
    FEEDBACK_VARIABLES feedback;
    ... etc
}MAIN_DATA;

static MAIN_DATA data;
```

This is cleaner as you can group related variables in their own struct, and you can have the global struct called data which you can then access like `data.gains.PropTerm`. This is better because you will have less separate global variables and will be less prone to errors. If you had a global variable called `i` and later you had a for loop that also used a variable `i` you can't say for certain which one would be used.